packages/labs/ssr/src/lib/declarative-style-dedupe.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {digestForTemplateResult} from '@lit-labs/ssr-client';
+import {CSSResult, CSSResultOrNative, TemplateResult} from 'lit';
+
+/**
+ * StyleDedupe is a utility that can be optionally passed to your
+ * SSR
+ */
+export class DeclarativeStyleDedupeUtility {
+  seenStyles = new Set<string>();
+  private hasEmittedDeclaration = false;
+
+  emitCustomElementDeclaration(): string {
+    if (this.hasEmittedDeclaration) {
+      return '';
+    }
+    this.hasEmittedDeclaration = true;
+    return `
+<script>
+
+class StyleModule extends HTMLElement {
+    static styleCache = new Map();
+
+    connectedCallback() {
+        // This attribute tells us that we expect the child styles
+        // to be cached.
+        if (this.hasAttribute("need-cached-styles")) {
+            return this.parsedCallback();
+        }
+
+        const obs = new MutationObserver(() => {
+            obs.disconnect();
+            this.parsedCallback();
+        });
+        obs.observe(this.parentNode, {childList: true, subtree: true});
+    }
+
+    parsedCallback() {
```

Comment:
I don't think you need this. Just put the style tag (not the text content) into the styleCache. By the time any other style-module requests the styles they will have necessarily been fully parsed.

Reply:
Yep – this simplifies the client a lot. Thank you! Will do this!